File: 프로그래머스/3/12927. 야근 지수/야근 지수.py
import heapq

def solution(n, works):
    totalPatigues = 0
    maxHeap = []
    
    for work in works:
        maxHeap.append(-work)
        totalPatigues += work**2
    heapq.heapify(maxHeap)
    
    for _ in range(n):
        # 가장 비용이 비싼 work를 꺼낸다
        work = heapq.heappop(maxHeap)
        
        # 최대힙으로 만들기 위해 -를 했으므로 다시 양수로 바꿈
        work = -work
        
        # 만약 현재 꺼낸 가장 비용이 비싼 work가 0이라면 모든 업무 처리한 것이니 0으로 리턴
        if work <= 0:
            return 0
        
        # 현재 작업했던 비용을 전체 피로도에서 감소
        totalPatigues -= (work) ** 2
        
        # 현재 작업을 1만큼 진행시킴
        work -= 1
        
        # 현재 비용만큼을 전체 피로도에 추가
        totalPatigues += (work) ** 2
        heapq.heappush(maxHeap, -work)

    
    
    return totalPatigues

File: 프로그래머스/3/12927. 야근 지수/야근 지수.js
class MaxHeap {
    constructor() {
        this.heap = [];
    }

    insert(value) {
        this.heap.push(value);
        this._bubbleUp();
    }

    extractMax() {
        if (this.heap.length === 0) return null;
        if (this.heap.length === 1) return this.heap.pop();
        const max = this.heap[0];
        this.heap[0] = this.heap.pop();
        this._sinkDown();
        return max;
    }

    _bubbleUp() {
        let idx = this.heap.length - 1;
        while (idx > 0) {
            let parentIdx = Math.floor((idx - 1) / 2);
            if (this.heap[parentIdx] >= this.heap[idx]) break;
            [this.heap[idx], this.heap[parentIdx]] = [this.heap[parentIdx], this.heap[idx]];
            idx = parentIdx;
        }
    }

    _sinkDown() {
        let idx = 0;
        const length = this.heap.length;
        const element = this.heap[0];

        while (true) {
            let leftChildIdx = 2 * idx + 1;
            let rightChildIdx = 2 * idx + 2;
            let leftChild, rightChild;
            let swap = null;

            if (leftChildIdx < length) {
                leftChild = this.heap[leftChildIdx];
                if (leftChild > element) {
                    swap = leftChildIdx;
                }
            }
            if (rightChildIdx < length) {
                rightChild = this.heap[rightChildIdx];
                if (
                    (swap === null && rightChild > element) ||
                    (swap !== null && rightChild > leftChild)
                ) {
                    swap = rightChildIdx;
                }
            }
            if (swap === null) break;
            [this.heap[idx], this.heap[swap]] = [this.heap[swap], this.heap[idx]];
            idx = swap;
        }
    }

    size() {
        return this.heap.length;
    }
}


function solution(n, works) {
    var answer = 0;
    
    const heap = new MaxHeap();
    
    for(const work of works){
        heap.insert(work)
    }
    
    
    for(let i=0; i<n; i++){
        const target = heap.extractMax()
        if(target <= 0)
            return 0
        heap.insert(target-1)
    }
    
    while (heap.size() > 0){
        const target = heap.extractMax()
        if(target >= 0){
            answer += target**2
        }
    }
    
    
    return answer;
}